File: src/knucleotide/knucleotide.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

type entry struct {
	key   string
	count int
}

var targets = []string{
	"GGT",
	"GGTA",
	"GGTATT",
	"GGTATTTTAATT",
	"GGTATTTTAATTTATAGT",
}

func encodeBase(ch byte) int {
	switch ch {
	case 'A':
		return 0
	case 'C':
		return 1
	case 'G':
		return 2
	case 'T':
		return 3
	default:
		return -1
	}
}

func decodeBase(code int) byte {
	return "ACGT"[code&3]
}

func readThreeSequence() []byte {
	scanner := bufio.NewScanner(os.Stdin)
	var sequence []byte
	inThree := false
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, ">") {
			if inThree {
				break
			}
			inThree = strings.HasPrefix(line, ">THREE")
			continue
		}
		if !inThree {
			continue
		}
		for i := 0; i < len(line); i++ {
			ch := line[i]
			if 'a' <= ch && ch <= 'z' {
				ch -= 'a' - 'A'
			}
			if encodeBase(ch) >= 0 {
				sequence = append(sequence, ch)
			}
		}
	}
	return sequence
}

func printFrequencies(sequence []byte, k int) {
	if len(sequence) < k {
		fmt.Println()
		return
	}
	bucketCount := 1 << (2 * k)
	mask := bucketCount - 1
	counts := make([]int, bucketCount)
	rolling := 0
	for i := 0; i < k; i++ {
		rolling = (rolling << 2) | encodeBase(sequence[i])
	}
	counts[rolling]++
	for i := k; i < len(sequence); i++ {
		rolling = ((rolling << 2) & mask) | encodeBase(sequence[i])
		counts[rolling]++
	}

	entries := make([]entry, 0, bucketCount)
	for code, count := range counts {
		if count == 0 {
			continue
		}
		key := make([]byte, k)
		value := code
		for i := k - 1; i >= 0; i-- {
			key[i] = decodeBase(value)
			value >>= 2
		}
		entries = append(entries, entry{key: string(key), count: count})
	}
	sort.Slice(entries, func(i, j int) bool {
		if entries[i].count != entries[j].count {
			return entries[i].count > entries[j].count
		}
		return entries[i].key < entries[j].key
	})

	total := float64(len(sequence) - k + 1)
	for _, entry := range entries {
		fmt.Printf("%s %.3f\n", entry.key, 100.0*float64(entry.count)/total)
	}
	fmt.Println()
}

func countFragment(sequence []byte, target string) int {
	fragment := []byte(target)
	count := 0
	for i := 0; i+len(fragment) <= len(sequence); i++ {
		match := true
		for j := 0; j < len(fragment); j++ {
			if sequence[i+j] != fragment[j] {
				match = false
				break
			}
		}
		if match {
			count++
		}
	}
	return count
}

func main() {
	sequence := readThreeSequence()
	printFrequencies(sequence, 1)
	printFrequencies(sequence, 2)
	for _, target := range targets {
		fmt.Printf("%d\t%s\n", countFragment(sequence, target), target)
	}
}

File: src/knucleotide/knucleotide.rs
use std::cmp::Ordering;
use std::io::{self, Read};

const TARGETS: [&str; 5] = [
    "GGT",
    "GGTA",
    "GGTATT",
    "GGTATTTTAATT",
    "GGTATTTTAATTTATAGT",
];

fn encode_base(ch: u8) -> usize {
    match ch {
        b'A' => 0,
        b'C' => 1,
        b'G' => 2,
        b'T' => 3,
        _ => usize::MAX,
    }
}

fn decode_base(code: usize) -> u8 {
    b"ACGT"[code & 3]
}

fn read_three_sequence() -> Vec<u8> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut in_three = false;
    let mut sequence = Vec::new();
    for line in input.lines() {
        if let Some(header) = line.strip_prefix('>') {
            if in_three {
                break;
            }
            in_three = header.starts_with("THREE");
            continue;
        }
        if !in_three {
            continue;
        }
        for mut byte in line.bytes() {
            byte.make_ascii_uppercase();
            if encode_base(byte) != usize::MAX {
                sequence.push(byte);
            }
        }
    }
    sequence
}

fn print_frequencies(sequence: &[u8], k: usize) {
    if sequence.len() < k {
        println!();
        return;
    }
    let bucket_count = 1usize << (2 * k);
    let mask = bucket_count - 1;
    let mut counts = vec![0usize; bucket_count];
    let mut rolling = 0usize;
    for &byte in sequence.iter().take(k) {
        rolling = (rolling << 2) | encode_base(byte);
    }
    counts[rolling] += 1;
    for &byte in sequence.iter().skip(k) {
        rolling = ((rolling << 2) & mask) | encode_base(byte);
        counts[rolling] += 1;
    }

    let mut entries: Vec<(Vec<u8>, usize)> = counts
        .into_iter()
        .enumerate()
        .filter(|(_, count)| *count > 0)
        .map(|(code, count)| {
            let mut key = vec![b'A'; k];
            let mut value = code;
            for index in (0..k).rev() {
                key[index] = decode_base(value);
                value >>= 2;
            }
            (key, count)
        })
        .collect();

    entries.sort_by(|a, b| match b.1.cmp(&a.1) {
        Ordering::Equal => a.0.cmp(&b.0),
        other => other,
    });

    let total = (sequence.len() - k + 1) as f64;
    for (key, count) in entries {
        println!(
            "{} {:.3}",
            String::from_utf8(key).unwrap(),
            100.0 * count as f64 / total
        );
    }
    println!();
}

fn count_fragment(sequence: &[u8], fragment: &[u8]) -> usize {
    if sequence.len() < fragment.len() {
        return 0;
    }
    sequence
        .windows(fragment.len())
        .filter(|window| *window == fragment)
        .count()
}

fn main() {
    let sequence = read_three_sequence();
    print_frequencies(&sequence, 1);
    print_frequencies(&sequence, 2);
    for target in TARGETS {
        println!("{}\t{}", count_fragment(&sequence, target.as_bytes()), target);
    }
}
